### ISSUE TITLE
Benchmarked Laya on an external look-alike set (68.8% best checkpoint) — corrections welcome

### ISSUE THREAD

**Issue by ihubanov** [NandhaKishorM/laya#135]
Hi — we maintain [Hunch](https://github.com/ihubanov/hunch), which gets typed decisions out of general LLMs by reading logprobs. Laya is the obvious "purpose-trained model instead" alternative, so we measured it on our benchmark and wrote the result up here: **[docs/COMPARISON-laya.md](https://github.com/ihubanov/hunch/blob/main/docs/COMPARISON-laya.md)**.

The benchmark is 240 fictional labelled pairs: "does NEW replace OLD's value for the same thing?" and "do A and B assert the same value?", deliberately full of restatements and same-value-different-thing traps. One `noul` question each, two variants (with and without definitions of yes/no). Script: [`bench/laya_compare.py`](https://github.com/ihubanov/hunch/blob/main/bench/laya_compare.py).

On one edge-class GPU, concurrency 1:

| Checkpoint | Accuracy @ p≥0.9 | AUROC | ECE | ms/question |
| --- | --- | --- | --- | --- |
| `laya-typed-decisions` | 68.8% | 0.759 | 0.207 | 62 |
| `laya` | 72.1% | 0.757 | 0.332 | 60 |
| `laya-multilingual` | 63.3% | 0.696 | 0.375 | 50 |

For context on the same pairs and GPU: a 9B general LLM reaches 90.8% at 141 ms, and a 27B one 99.6%. A 0.6B general LLM manages 35.8%, so **purpose-training clearly beats raw generality at that size** — Laya's 421M is far ahead of a 0.6B LLM.

We got one thing wrong ourselves and fixed it: we first appended the yes/no definitions to `instructions` as prose, because the `system_one` docstring lists `criteria` only for `choice` and `score`. Using Laya's native `noul` criteria was worth about 3 points. That's written up in the document.

Four questions, in case we're still holding it wrong:

1. Is `criteria` with `{"true": ..., "false": ...}` the intended interface for `noul`? The docstring omits it, though `_to_internal` passes it through and your presets use it.
2. Should this workload go through `Router` / `presets` / `task="typed_decisions"` rather than `Agent.predict` directly?
3. On load, `laya` and `laya-typed-decisions` print a clamp warning for temperatures outside `[0.5, 5]` (`choice:11+`). Expected?
4. Any recommended formatting when the state is a dict of two short strings?

Happy to rerun anything and correct the write-up. The benchmark is deliberately adversarial and is only one task, which the document says plainly.

**Comment by chilkotiKartik** [comment]
Thanks for running and publishing this comparison. Seeing 421M non-autoregressive decision models hold up so well against 0.6B-9B general LLMs on adversarial pair matching is really encouraging.

To answer your four questions specifically:

### 1. `criteria` with `{"true": ..., "false": ...}` on `noul` questions
Yes, `criteria={"true": "...", "false": "..."}` is the intended interface for `noul` (binary) decisions. 
Under the hood, `_to_internal()` formats the criteria keys as candidate decision boundaries and tokenizes them into the cross-attention head alongside the prompt and state. While earlier docstrings only highlighted criteria examples on `choice` (categorical) and `score` (ordinal), the underlying scoring head was trained with explicit boundary definitions. Providing explicit criteria helps ground edge cases and gives the ~3 point boost you observed.

### 2. `Router` / `presets` vs direct `Agent.predict`
For homogeneous benchmarking in a single language (like English assertion pairs), calling `Agent.predict(state, questions)` with the specific checkpoint (e.g. `laya-typed-decisions` or `laya`) is optimal because it avoids the lightweight language detection overhead on every request. 
`Router` is primarily intended for production deployments with mixed-language traffic or multi-task dispatch where you want automatic selection between the base model, `laya-multilingual`, and specialized task heads.

### 3. Logit calibration temperature clamp warnings
The warning you saw (`clamp warning for temperatures outside [0.5, 5]`) comes from the Platt scaling / temperature calibration layer in `laya/common.py`. When loading older exported weights where certain decision heads have very large uncalibrated logit ranges, the temperature optimizer clamps extreme values into `[0.5, 5.0]` to prevent overconfidence or vanishing gradients. This is harmless at inference time, and PR #142 addresses making those temperature reporting bounds cleaner without logging false-positive warnings.

### 4. Dict formatting for multi-variable state
For two-variable comparison (like `NEW` vs `OLD` or pair assertions), structuring the state dictionary with clean semantic keys works best:

```python
state = {
    "old_assertion": "Invoice #1024 amount is $450.00",
    "new_assertion": "Invoice #1024 amount was updated to $450.00"
}

questions = {
    "is_replacement": {
        "type": "noul",
        "question": "Does the new assertion modify the value of the old assertion?",
        "criteria": {
            "true": "The new assertion changes or replaces the specific value stated previously.",
            "false": "The new assertion restates the same value or refers to an entirely different entity."
        }
    }
}
```

Passing structured dicts allows the tokenizer to format the key-value pairs cleanly into the state buffer without losing field boundaries.

**Comment by ihubanov** [comment]
Thanks — that was useful, and I tested the one suggestion that could move the numbers rather than just taking it on trust.

**Context-key naming (your point 4).** Same 240 pairs, same machine, bare `{"old", "new"}` vs descriptive `{"old_assertion", "new_assertion"}`:

| Checkpoint | Accuracy @ p≥0.9 | Vague | AUROC | ECE |
| --- | --- | --- | --- | --- |
| `laya-typed-decisions`, bare | 68.8% | 66.7% | 0.759 | 0.207 |
| `laya-typed-decisions`, descriptive | 68.8% | 66.7% | 0.734 | 0.204 |
| `laya`, bare | 72.1% | 72.5% | 0.757 | 0.331 |
| `laya`, descriptive | 67.5% | 72.1% | 0.691 | 0.304 |

No help on this task: unchanged on the purpose-built checkpoint (ranking slightly worse), and clearly worse on the English one. Both deterministic, no errors. It's now a flag in our harness (`bench/laya_compare.py --semantic-keys`) so either variant reproduces, and the result is in our write-up.

Your other three answers matched what we found, and I've noted them: `criteria` on `noul` is intended (it was worth ~3 points to us), `Agent.predict` is the right call for single-language benchmarking, and the temperature clamp is a calibration-layer artefact rather than something we caused.

Two small things back, offered as observations rather than complaints:

1. The snippet in your point 4 passes `"question": ...` inside the check, but `Agent.system_one` reads `qdef["instructions"]`, so that example raises `KeyError` as written. Worth fixing if it gets copied into docs.
2. Since `criteria` is intended for `noul`, the `system_one` docstring is the one place that says otherwise — it lists `criteria` only for `choice` and `score`. That's exactly what led us to format it as prose first and under-measure the model by ~3 points. A one-line docstring change would prevent the next person doing the same.

For completeness, the gap we're reporting is about 21 points to our qualification gate, and the largest movement any formatting change produced was 4.6 points — so we don't think formatting explains it. The benchmark is deliberately adversarial about near-identical meaning and is only one task; our write-up says so. If there's a configuration you'd expect to do better, we're happy to run it.

**Comment by NandhaKishorM** [comment]
Thanks for running this carefully and publishing it, including the part you corrected yourselves.

Your two observations are both right:

1. The `system_one` docstring still lists `criteria` only for `choice` and `score`. `{"true": ..., "false": ...}` on `noul` is intended and supported, and the docstring is the one place that says otherwise. That will be fixed.
2. In the snippet above, `"question"` should be `"instructions"`: `system_one` reads `qdef["instructions"]`, so that example raises `KeyError` as written.

One thing that may matter a lot for your benchmark: it is all `noul`, and #156 shows the current checkpoints are biased by the rendered `false:`/`true:` option labels, which `noul` hardcodes. On some states the model follows the label word rather than the content, and it does so at high confidence. A cheap check is to run the same 240 pairs as a two-option `choice` with neutral keys (`A`/`B`) and your yes/no definitions as the descriptions. If accuracy moves, your write-up is measuring the label bias as much as the task.

That said, a 21-point gap to your gate on a deliberately adversarial near-duplicate task is plausible for a 421M model, and your framing of it as one task is fair.

**Comment by ihubanov** [comment]
You were right, and it moved a lot. Thank you for pointing at the control rather than arguing with the numbers — our write-up understated Laya and is now corrected.

Same 240 pairs, same machine, `noul` vs a two-option `choice` with neutral `A`/`B` keys and the same definitions as the option descriptions, run in both key orders and averaged so a position preference cancels:

| Checkpoint | Asked as | Accuracy @ p≥0.9 | AUROC | Brier | ECE |
| --- | --- | --- | --- | --- | --- |
| `laya` (English) | `noul` | 72.1% | 0.757 | 0.315 | 0.332 |
| `laya` (English) | **neutral choice** | **81.2%** | **0.953** | 0.215 | 0.301 |
| `laya-typed-decisions` | `noul` | 68.8% | 0.759 | 0.224 | 0.207 |
| `laya-typed-decisions` | **neutral choice** | **73.8%** | **0.904** | 0.189 | 0.236 |

The line we published — "barely above a constant-no baseline, AUROC 0.759, only weak signal" — is wrong and has been removed. Through `noul` the label words were suppressing the model's discrimination, not merely costing it a few points. AUROC 0.953 on the English checkpoint is a model that ranks these look-alikes well; for comparison a 9B general LLM scores 0.996 on the same pairs, in a model 20× larger.

Two further things from the re-run that may be useful to #156:

1. **Calibration did not improve with neutral labels** — ECE 0.332 → 0.301 and 0.207 → 0.236 (slightly worse on the purpose-built checkpoint). So on our set the label effect looks like it suppresses *discrimination* while leaving the probability scale roughly as miscalibrated as before. Our honest headline is now "good ranker, probabilities you would have to calibrate yourself" rather than "it fails".
2. **The direction of the `criteria` effect flips with the framing.** Under `noul`, adding yes/no definitions made the English checkpoint *worse* (72.5% → 72.1%); as a neutral choice the same definitions are worth +14.5 points (66.7% → 81.2%). If a `noul` fix lands, that is a cheap regression check.

One limit on our claim, stated in the write-up: we verified that the control changes the result on our set, not that #156's mechanism is what causes it.

The comparison is updated with both corrections we have now published against ourselves — the first was our own bug (prose instead of native `criteria`, ~3 points), this one is yours: https://github.com/ihubanov/hunch/blob/main/docs/COMPARISON-laya.md . `bench/laya_compare.py --as-choice` reproduces it.
